### ISSUE TITLE
Defunct processes still happen

### ISSUE THREAD

**Issue by gerryd** [krallin/tini#42]
Hi!

I'm probably doing something wrong, or at least something I'm not supposed to do, but I'm still seeing defunct processes when using tini.

The idea is to run a container with Apache and MySQL. Before both of these start, an NFS share needs to be mounted, containing the database files and files required by the application itself. MySQL is configured to read its database files from that share, as is the application.

Starting everything up is not a problem, daemons start, tests can be run, but shutting the container down is a problem. That's when I stumbled across tini, which seemed would solve the issues.

No such luck, I'm still seeing defunct processes:

```
    root       507  4.9  0.3 665036 56756 ?        Ssl  16:11   0:28 /usr/bin/docker daemon -H fd://
    root       537  0.0  0.0 232400 14768 ?        Ssl  16:11   0:00  \_ docker-containerd -l /var/run/docker/libcontainerd/docker-containerd.sock --runtime docker-runc --start-timeout 2m
    root     10726  0.0  0.0 274592  3360 ?        Sl   16:16   0:00      \_ docker-containerd-shim 88ae86ccaee1221bc6ee15afde41037bd5461caf7f9c51241ee855f671bfda0e /var/run/docker/libcontainerd/88ae86ccaee1221bc6
    root     10742  0.0  0.0      0     0 ?        Ss   16:16   0:00          \_ [tini]
    sshd     10765  0.2  0.0      0     0 ?        Zl   16:16   0:00              \_ [mysqld] <defunct>
```

Or like so:

```
root      1511  0.2  0.2 719456 39432 ?        Ssl  16:39   0:01 /usr/bin/docker daemon -H fd://
root      1516  0.0  0.0 362416 13456 ?        Ssl  16:39   0:00  \_ docker-containerd -l /var/run/docker/libcontainerd/docker-containerd.sock --runtime docker-runc --start-timeout 2m
root      1848  0.0  0.0 266652  3084 ?        Sl   16:39   0:00      \_ docker-containerd-shim d95d83caa0b4e23cd81a938630fbcd3738208a59311221683575833bc02d2f06 /var/run/docker/libcontainerd/d95d83caa0b4e23cd81a938630fbcd3738208a59311221
root      1863  0.0  0.0      0     0 ?        Ss   16:39   0:00          \_ [tini]
www-data  1896  0.0  0.0      0     0 ?        D    16:39   0:00              \_ [apache2]
```

I'm all out of ideas.

This is my Dockerfile:

```
FROM debian:jessie

STOPSIGNAL 9

# Install packages
ENV DEBIAN_FRONTEND noninteractive

# Configure Percona repository
RUN echo "deb http://repo.percona.com/apt jessie main" >> /etc/apt/sources.list.d/percona.list
RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 0x1C4CBDCDCD2EFD2A

# Update and install packages
RUN apt-get update && apt-get -y install curl nfs-common git apache2 libapache2-mod-php5 percona-server-server-5.6 php-apc php-pear php5-apcu php5-cli php5-common php5-curl php5-gd php5-intl php5-json php5-mcrypt php5-mysql php5-readline php5-tidy php5-xmlrpc && apt-get install -y --no-install-recommends psutils

# Install composer (please find a better way :'()
RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/bin --filename=composer

# Add image configuration and scripts
ADD config/mysql.cnf /etc/mysql/conf.d/mysql.cnf

ADD scripts/entrypoint.sh /opt/entrypoint.sh
RUN chmod 755 /opt/entrypoint.sh

# Apache configuration
ADD /config/apache-vhost.conf /etc/apache2/sites-available/000-default.conf
RUN a2enmod rewrite

ADD https://github.com/krallin/tini/releases/download/v0.9.0/tini-static /tini
RUN chmod 755 /tini

# PHP environment variables
ENV PHP_UPLOAD_MAX_FILESIZE 10M
ENV PHP_POST_MAX_SIZE 10M

EXPOSE 80 443

ENTRYPOINT ["/tini", "--", "/opt/entrypoint.sh"]
```

And this is the entrypoint.sh script:

```
#!/usr/bin/env bash
set -x

# start rpcbind
rpcbind -f &

# mount data
mkdir -p /mnt/data/
mount -o nfsvers=3 10.1.2.3:/mnt/pool0/snapshot/1234 /mnt/data/

# start mysql
mysqld &

# start apache
source /etc/apache2/envvars
apache2 -D FOREGROUND &

# launch a shell
/bin/bash
```

**Comment by krallin** [comment]
I'm not entirely sure I fully understand what you're doing yet, and reproducing this might be a little tricky for me considering the NFS set up. That being said, could you please run tini with the `-vvv` flag, stop your container, and post its output here? That would at least give me a better picture of what it it's that you're actually doing here ;)

To be fair, I'm surprised Tini isn't even exiting immediately when bash exits (which I assume happens since it's not in the process lists you're showing).

Thanks!

**Comment by gerryd** [comment]
Thanks for looking into this so quickly. :)

This is the output I'm getting:

```
$ ps faxu
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.2  0.0   1116     4 ?        Ss   15:25   0:00 /tini -vvv -- /opt/entrypoint.sh sh -c if [ -x /usr/local/bin/bash ]; then .exec /usr/local/bin/bash  elif [ -x /usr/bin/bash ]; then .exec /usr/bin/bash  elif [ -x /bin/bash ]; then .exec /bin/bash  elif [ -x /usr/local/bin/sh ]; then .exec /usr/local/bin/sh  elif [ -x /usr/bin/sh ]; then .exec /usr/bin/sh  elif [ -x /bin/sh ]; then .exec /bin/sh  else .echo shell not found .exit 1 fi  
root         7  0.0  0.0  20064  2860 ?        S    15:25   0:00 bash /opt/entrypoint.sh sh -c if [ -x /usr/local/bin/bash ]; then .exec /usr/local/bin/bash  elif [ -x /usr/bin/bash ]; then .exec /usr/bin/bash  elif [ -x /bin/bash ]; then .exec /bin/bash  elif [ -x /usr/local/bin/sh ]; then .exec /usr/local/bin/sh  elif [ -x /usr/bin/sh ]; then .exec /usr/bin/sh  elif [ -x /bin/sh ]; then .exec /bin/sh  else .echo shell not found .exit 1 fi  
root         8  0.0  0.0  37080  2872 ?        S    15:25   0:00  \_ rpcbind -f
mysql       15  8.7  0.0      0     0 ?        Zl   15:25   0:00  \_ [mysqld] <defunct>
root        16  1.4  0.1 402108 31948 ?        S    15:25   0:00  \_ apache2 -D FOREGROUND
www-data    19  0.0  0.0 402140  9544 ?        S    15:25   0:00  |   \_ apache2 -D FOREGROUND
www-data    20  0.0  0.0 402140  9544 ?        S    15:25   0:00  |   \_ apache2 -D FOREGROUND
www-data    21  0.0  0.0 402140  9544 ?        S    15:25   0:00  |   \_ apache2 -D FOREGROUND
www-data    22  0.0  0.0 402140  9544 ?        S    15:25   0:00  |   \_ apache2 -D FOREGROUND
www-data    23  0.0  0.0 402140  9544 ?        S    15:25   0:00  |   \_ apache2 -D FOREGROUND
root        49  0.0  0.0  20064  2780 ?        S    15:26   0:00  \_ /bin/bash
root        51  0.0  0.0  20072  2360 ?        S    15:26   0:00      \_ /bin/bash
root        54  0.0  0.0  17496  2012 ?        R    15:26   0:00          \_ ps faxu
statd       14  0.5  0.3  92576 55948 ?        Ss   15:25   0:00 rpc.statd --no-notify
/opt/entrypoint.sh: line 28:    15 Killed                  mysqld
[INFO  tini (1)] Spawned child process '/opt/entrypoint.sh' with pid '7'
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[TRACE tini (1)] No child to reap
[DEBUG tini (1)] Received SIGCHLD
[DEBUG tini (1)] Reaped child with pid: '7'
[INFO  tini (1)] Main child exited normally (with status '0')
[TRACE tini (1)] No child to reap
[TRACE tini (1)] Exiting: child has exited
```

The MySQL process is being killed from within entrypoint.sh, which allows me to reproduce the issue consistently. Notice the defunct MySQL process in the processlist.

**Comment by krallin** [comment]
@gerryd It looks like MySQL (PID 15) is a child of bash (PID 7), so bash is getting the `SIGCHLD`, Tini isn't. Tini will only receive `SIGCHLD` when / if MySQL's parent (which is bash) exits and MySQL is re-parented to Tini.

The solution to your problem is to let MySQL and Apache daemonize rather than merely launch them in the background (which you're doing by disabling daemonization and using `&` to send them to the background).

Note: it'd be very incorrect (and probably impossible, though I haven't checked lately) for Tini to reap processes it doesn't own. That would confuse their rightful owners.

Cheers,

**Comment by krallin** [comment]
Can you clarify what you mean by "after the issue occurs:"? How do you trigger said issue? With `docker stop`? if so can you send the output, and perhaps check your docker logs as well.

I'll try to repro your issue without the NFS parts and see whether that works on my end.

Thanks!

**Comment by yosifkit** [comment]
It looks like you are running bash as the final process of the container?  What I see is that tini has a child of bash and inside that you run your services and end with a second bash process that you look at interactively.  In this case tini will only forward signals to its immediate children (ie, just bash) and so when bash exits, tini then exits, killing all other processes.

Rather than running multiple processes in the background, you should either split them into multiple containers (and use something like compose to bring them all up), or use a supervisor like http://supervisord.org/ to manage the multiple processes. Currently if you had an issue with mysql or nfs dying, the container would not automatically bring it back up.

**Comment by gerryd** [comment]
@krallin What I meant was the first process list I pasted is when the container is running, so you get an idea of what is being run and how the processes are linked. The second one ("after the issue occurs") is the processlist showing the defunct process, which happens when the container gets stopped.
The Docker container is being run as part of a GitLab CI process, so I'm not entirely sure how the CI runner shuts down the container, but I would assume it calls `docker stop`, indeed. The strange thing is, and I just noticed that, I can consistently reproduce it when it is being controlled via the CI runner, but not when I'm running it myself from the commandline. Something must be different between the two methods.
As for logs, I don't seem to be able to find relevant logs anywhere. I would think Docker would log the commands issues by the GitLab CI runner (or via the `docker` command) somewhere, but I don't see it, and the internet isn't being helpful. Any pointer as to what you are looking for, exactly?

@yosifkit The previous iteration of this Docker container was using supervisord, and had the same issue. Most of the time, it was the `mysqld` process that would go defunct.
Splitting up into multiple containers would indeed be the way to go, but as this is part of a GitLab CI flow, I'm rather limited in my options. I need a fresh snapshot of the data & databases (supplied via the NFS mount), and I need a MySQL and Apache process. The requirement of getting and mounting the snapshot is what makes things complicated, I don't think there is a way to do that (in combination with GitLab CI) without spawning a shell script before starting the daemons.

If this can't be fixed (I agree I'm probably abusing Docker & tini), I'll have to resort to using the GitLab CI shell executor instead of the Docker one, allowing for more control over the containers. I could then use Docker volumes to get the data snapshots into the containers, and could then even run them all separately. I would prefer not to, I like to avoid reinventing the wheel and using what is there already, if at all possible.

**Comment by krallin** [comment]
@gerryd,

It'd be worthwhile if you could retrieve the Docker daemon logs corresponding to your container being terminated. It looks like your container should receive `SIGKILL` (as per your `STOPSIGNAL`), and no matter what, Docker should *eventually* send a `SIGKILL`, which should tear down the entire container. If the container does in fact receive a `SIGKILL`, then there isn't much Tini can do for you. Once Tini receives a `SIGKILL`, it'll *never* run again. 

That being said, in your example, it looks like Tini might have exited normally already anyway, which should result in your entire container being torn down since Tini is running as PID 1. However, that's not cleanup Tini itself should do: at that stage, it's up to the Kernel to do the cleanup (since Tini itself has exited).

I have certainly seen some cases where processes can't be destroyed immediately because they had some IO pending, etc. which tends to happen especially when you're using somewhat unreliable IO, like NFS. The fact that your MySQL container is using CPU despite being dead suggests something like this might be up. There are bunch of Docker issues you could look at for more information (just search for "Docker cannot kill container").

---

Does the condition persist for a while after you terminate the container, or does it eventually go away?

**Comment by krallin** [comment]
Hey there; haven't heard back from you in a while. is this still an issue for you?

**Comment by gerryd** [comment]
Hi!

Sorry for going silent. I tried a bunch more things, before I eventually had to give up due to a lack of time and  switch the GitLab CI runner to `shell`, allowing me to create dedicated containers per service. From a GitLab CI point of view, this is a bit of a clumsy workaround, but it makes sense from a Docker perspective.

I still don't understand fully what was going on and why it didn't work the way I wanted it to. I suggest to close this issue for now, as unfortunately I don't have time to dig into this further right now. If another project comes along, I'll give it another go. I'll open a new issue if need be at that point.

Thanks for your input, if anything, it shed at least some light on what could be the problem!

**Comment by krallin** [comment]
Hey @gerryd, no worries; thanks for getting back to me 😄 

I'm sorry I haven't been able to reproduce this. I'm still skeptical the issue is in Tini itself, but if anything new comes up / if you end up looking at this again, please do comment here and I'll reopen!

Thanks!
